server/src/core/agent/cache.py

```diff
@@ -1,21 +1,26 @@
-from core.common.encode import encode_to_base64, decode_from_base64
-from core.common.cache import CacheMemory
+# from core.common.encode import encode_to_base64, decode_from_base64
+from core.common import conn
 
 
+# TODO: Why were there used base64 encodings?
+#       Experiments show that it receives binary strings as arguments and works fine.
 class PredictCache:
     def __init__(self):
-        self.cache = CacheMemory(30)
+        # self.cache = CacheMemory(30)  # TODO: Limit the number of cached entries?
+        self.redis = conn.get_redis_instance()  # TODO: Rewrite with dependency injection.
 
-    def put(self, key, value: str):
-        k = encode_to_base64(key)
-        v = encode_to_base64(value)
-        self.cache.put(k, v)
+    async def put(self, key, value: str):
+        # k = encode_to_base64(key)
+        # v = encode_to_base64(value)
+        await self.redis.hset("predict_cache:" + key, "value", value)
 
-    def get(self, key):
-        k = encode_to_base64(key)
-        data = self.cache.get(k)
-        if data is not None:
-            return decode_from_base64(data)
+    async def get(self, key):
+        # k = encode_to_base64(key)
+        data = await self.redis.hget("predict_cache:" + key, "value")
+        return data
+        # if data is not None:
+        #     return decode_from_base64(data)
 
-    def exists(self, k) -> bool:
-        return self.get(k) is not None
+    async def exists(self, key) -> bool:
+        # k = encode_to_base64(key)
```

Comment:
it is a good practice to hash keys. Keep the encode for the key

Reply:
Hashing is different than base64 encode. Using base64 seems having no sense. Should I SHA or like this hash the key? (I am sure that Redis does hashing the keys internally, but if we hash a long key, it takes less disk space.)

Reply:
You know what I meant. No hashing. 64 Encoding is enough. Hashing algos are very costly. Keep it simple

Reply:
I see no reason to use base64 encoding. It only somehow decreases performance and also complicates debugging (because the text is encoded).

If you have some reason to use base64, please tell it to me.

Reply:
Raw text from users as keys are not recommend. 

You are being picky here. It is not a big deal either way. Let's move on with base64

Reply:
> Raw text from users as keys are not recommend.

It seems that you confuse search keys at one side and API keys and hashes at the other side.

I know the rule that raw strings as API keys are not recommended. I don't know the rule not to use raw strings for search keys.

Moreover, base64 is not an encryption and therefore does not influence security.

I write the above to be sure that both of us don't misunderstand.

Reply:
There is no misunderstanding, nobody talked about security.

It is a bad practice to store raw data from the users as key. 

You might have to access the data from a terminal or any interface that might not deal well with especial chars. Remember that the input from the user can be anything in any length. 

I am not discussing this anymore since that is the basics of software development

Reply:
OK, thank you for explaining why base64 is desirable. I've added back base64 for keys, as you requested.

Reply:
I think the base64 encoding is fine, it's not encryption of course. Seems this thread was a bit of a miscommunication only. 

I do have a question more so than an 'issue' but out of interest @vporton what is the preference for the async function changes ?

Reply:
I don't understand what means "the preference for the async function changes".

If you ask, why I added `async` to several functions. This is because they now call other async functions and can't be sync syntaxically.